### PR TITLE
feat: finish code for background flat storage creation

### PR DESCRIPTION
**Pull request by Longarithm**
Implement two remaining steps for background flat storage creation:

### Fetching state

We split the state into several parts and fetch them one by one. Number of parts is based on the state size limit, which I set to 10 MiB. Then, fetching is executed in several steps so that we could save intermediate results. Currently each step includes 20 state parts. They are fetched using threads of a separate rayon pool, which size is limited by 4 threads so that it doesn't affect block processing.

Here I also introduce `FetchingStateStatus` which defines the current progress. It makes testing more convenient: for lightweight tests it is enough to fetch only one part, alhough on production we need thousands of parts.

After state for a shard is fully fetched, we start catching up - which means that we move flat storage head forward and apply all saved deltas, limiting it by 50 blocks at once.

After we fully caught up, we finally create flat storage state.

## Testing

* Complete scenario in `test_flat_storage_creation`.
* https://buildkite.com/nearprotocol/nearcore-flat-state/builds/108

### CODE REVIEW

**Review by jakmeier — COMMENTED** [review]
Looks solid to me, great to see this progress!

@mzhangmzz I think you should also take a look before we merge, feels like your expertise around the chain and flat state are required here.

@Longarithm Do you plan to add tests to this PR or will it be a follow-up? I would tend towards a separate follow-up PR but it's up to you.

File: chain/chain/src/flat_storage_creator.rs
Comment (Longarithm) [reply chain/chain/src/flat_storage_creator.rs:270]: Yeah, it's correct, I look at this as the way to avoid code duplication. Added a comment above.

File: chain/chain/src/flat_storage_creator.rs
Comment (Longarithm) [reply chain/chain/src/flat_storage_creator.rs:156]: Done, I also created issue for more general fix recently https://github.com/near/nearcore/issues/8057

**Comment by Longarithm** [comment]
> Do you plan to add tests to this PR or will it be a follow-up? I would tend towards a separate follow-up PR but it's up to you.

Yeah, they will be in follow-up PRs.

File: chain/chain/src/flat_storage_creator.rs
Comment (jakmeier) [reply chain/chain/src/flat_storage_creator.rs:270]: Thanks, yeah it makes perfect sense to do it this way. I just wanted to make sure I am not tricking myself into thinking I understand when maybe I don't understand ^^

File: chain/chain/src/flat_storage_creator.rs
Comment (mzhangmzz) [inline chain/chain/src/flat_storage_creator.rs:70]: Could you add a comment for this function? It's a bit hard to figure out what it does.

File: chain/chain/src/flat_storage_creator.rs
Comment (mzhangmzz) [inline chain/chain/src/flat_storage_creator.rs:134]: Could you add a comment to briefly describe what this function does?

File: chain/chain/src/flat_storage_creator.rs
Comment (mzhangmzz) [inline chain/chain/src/flat_storage_creator.rs:38]: Add comments explain why it is an Option.

File: chain/chain/src/flat_storage_creator.rs
Comment (mzhangmzz) [inline chain/chain/src/flat_storage_creator.rs:26]: Maybe add a paragraph at the beginning of this file to describe on a high level how flat state is migrated, what the steps are, and how FlatStateCreator is used in the code. I know that you already have comments throughout the code, but I think it is still valuable to have an overview to link everything together and help the readers to have a general picture before they dive into the code.

File: chain/chain/src/flat_storage_creator.rs
Comment (mzhangmzz) [inline chain/chain/src/flat_storage_creator.rs:292]: Make a constant for the number 50

File: chain/chain/src/flat_storage_creator.rs
Comment (mzhangmzz) [inline chain/chain/src/flat_storage_creator.rs:297]: Let's compare the hash instead of the height here. I think technically height is ok, since we are moving forward on the canonical chain, but just to be safe.

File: chain/chain/src/flat_storage_creator.rs
Comment (mzhangmzz) [inline chain/chain/src/flat_storage_creator.rs:306]: is it possible that old_flat_head == flat_head == final_head?

File: chain/chain/src/flat_storage_creator.rs
Comment (mzhangmzz) [inline chain/chain/src/flat_storage_creator.rs:315]: Same here, can we compare hash instead of height?

File: chain/chain/src/flat_storage_creator.rs
Comment (mzhangmzz) [inline chain/chain/src/flat_storage_creator.rs:319]: Is this debug statement intended? Should it print something like flat storage done?

File: chain/chain/src/flat_storage_creator.rs
Comment (mzhangmzz) [inline chain/chain/src/flat_storage_creator.rs:321]: nit: as a general practice, don't put something more than an evaluation statement inside assert. Let's move `self.runtime_adapter.try_create_flat_storage_state_for_shard` outside the assert and create a variable for the assert.

File: chain/chain/src/flat_storage_creator.rs
Comment (mzhangmzz) [inline chain/chain/src/flat_storage_creator.rs:193]: Should we store flat head here with a different prefix, maybe something like `set_temporary_flat_head` to distinguish it from when flat head is actually set and the flat state is ready to use? This way, there is no way the code can accidentally create a flat storage if flat state is not ready.

**Review by mzhangmzz — APPROVED** [review]
The PR looks great! Thanks for the awesome work and sorry for the super late review. 

Most of my comments are about adding more comments :) Let's try to make the code easier to read for other people and the future us. I approved it to unblock, but please address the comments.

File: chain/chain/src/flat_storage_creator.rs
Comment (Longarithm) [reply chain/chain/src/flat_storage_creator.rs:306]: In reality - no, because final head should move forward at least once during `FetchingState` step.
And even if it is the case, it is fine to wait on this step until final head moves forward.

File: chain/chain/src/flat_storage_creator.rs
Comment (Longarithm) [reply chain/chain/src/flat_storage_creator.rs:319]: Yeah, it is intended. 
I agree that "Flat storage creation done" sounds better. Also changed `debug!` to `info!` to places where major part of work is finished so users will be aware of status but there won't be much spam in logs.

File: chain/chain/src/flat_storage_creator.rs
Comment (Longarithm) [reply chain/chain/src/flat_storage_creator.rs:193]: We probably should. Let's do it separately because this change is already quite big and write some unit test checking that FS can't be created accidentally.

**Pull request merged**
